feat: Seamless calendar schedules

caseycharleston opened the pull request:
- feat: schedule list item vertical dots dropdown
- feat: handling duplicate schedules with plus button and duplicate button
- feat: renaming schedules to existing schedule names handled by duplicate name handler
- fix: deleting schedules with indices before the active index no longer moves active index down one
- feat: prompts for deleting schedules and attempting to delete active schedule

~

- ~~chore: vertical dots menu is misaligned since other component changes~~
- ~~missing: prompts for deleting schedules and attempting to delete active schedule not implemented~~
- ~~issue: pressing Rename in the dropdown menu does not properly focus on the schedule name~~

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)
<!-- Reviewable:end -->

Samathingamajig commented on src/views/components/common/ScheduleListItem/ScheduleListItem.tsx:
```diff
@@ -26,6 +32,35 @@ export default function ScheduleListItem({ schedule, dragHandleProps, onClick }:
+    // eslint-disable-next-line react/no-unescaped-entities
+    const deleteContent = <Text>Deleting "{schedule.name}" is permanent and will delete its related courses.</Text>;
+    const deleteChildren = [
+        <Button
+            key='yes'
+            variant='single'
+            color='ut-burntorange'
+            onClick={() => {
+                deleteSchedule(schedule.id);
+                setShowDeletePrompt(false);
+            }}
+        >
+            Yes
+        </Button>,
+        <Button key='no' variant='single' color='ut-burntorange' onClick={() => setShowDeletePrompt(false)}>
+            No
+        </Button>,
```
Follow the proper coloring and wording from the Figma

![image](https://github.com/Longhorn-Developers/UT-Registration-Plus/assets/29130894/c992cdcc-1850-4509-9b5c-b40365df456a)

caseycharleston replied:
- Fixed floating in commit 330c38c059c34c48e48d327c26d637bb814fa4c3
- Gonna go ahead and remove the dialog component until it is fully finished per Ellie's request

Samathingamajig commented on src/pages/background/lib/renameSchedule.ts:
```diff
@@ -12,9 +14,9 @@ export default async function renameSchedule(scheduleId: string, newName: string
-    //     return `Schedule ${newName} already exists`;
-    // }
+
+    // eslint-disable-next-line no-param-reassign
+    newName = await handleDuplicate(newName);
```
I believe making a new variable after the duplicate handling should be preferred to reassigning the parameter and disabling eslint

caseycharleston replied:
Done!

Samathingamajig commented on src/pages/background/lib/createSchedule.ts:
```diff
@@ -1,16 +1,19 @@
-    // }
+    // Duplicate schedule found, we need to append a number to the end of the schedule name
+
+    // eslint-disable-next-line no-param-reassign
+    scheduleName = await handleDuplicate(scheduleName);
```
I believe making a new variable after the duplicate handling should be preferred to reassigning the parameter and disabling eslint

caseycharleston replied:
Done! Will clean up the es lint disable in a future commit

Samathingamajig commented on src/pages/background/lib/handleDuplicate.ts:
```diff
@@ -0,0 +1,33 @@
+    const regex = / \((\d+)\)[^(]*$/;
+    let index = 1;
+
+    // Trim ending number
+    let baseName = scheduleName.replace(regex, '');
+    let newName = `${baseName}`;
```
`baseName` can be `const`, and we don't need to wrap something that's ready a string in a template literal
```suggestion
    // Trim ending number
    const baseName = scheduleName.replace(regex, '');
    let newName = baseName;
```

caseycharleston replied:
Done!

Samathingamajig commented on src/views/components/common/ScheduleListItem/ScheduleListItem.tsx:
```diff
@@ -26,6 +32,35 @@ export default function ScheduleListItem({ schedule, dragHandleProps, onClick }:
     const [editorValue, setEditorValue] = useState(schedule.name);
+    const [showDeletePrompt, setShowDeletePrompt] = React.useState(false);
+    const [showActiveDeletePrompt, setShowActiveDeletePrompt] = React.useState(false);
+    const deleteTitle = <Text>Are you sure?</Text>;
+    // eslint-disable-next-line react/no-unescaped-entities
+    const deleteContent = <Text>Deleting "{schedule.name}" is permanent and will delete its related courses.</Text>;
```
The title should be formatted as a title. The name should be colored, not in quotes
![image](https://github.com/Longhorn-Developers/UT-Registration-Plus/assets/29130894/8d90343d-6860-4676-9d98-2696efe76333)

(also, if it was in quotes, prefer the entity (`&quot;`) to disabling eslint)

caseycharleston replied:
- Removing dialog usage until the component is finalized per Ellie's request

Samathingamajig commented on src/views/components/common/ScheduleListItem/ScheduleListItem.tsx:
```diff
@@ -26,6 +32,35 @@ export default function ScheduleListItem({ schedule, dragHandleProps, onClick }:
+        </Button>,
+    ];
+
+    const activeDeleteTitle = <Text>Invalid action!</Text>;
+    const activeDeleteContent = <Text>Deleting the active schedule is disallowed.</Text>;
```
like above, the title should be formatted

caseycharleston replied:
Removing dialog usage until the component is finalized per Ellie's request

Samathingamajig reviewed (changes requested):
overall, good

the experience within the extension popup is not pleasant, i need to scroll in that tiny box to see all 3 of the options

Razboy20 reviewed (changes requested):
+awaiting-changes

I only put comments on a couple things in this review, however keep in mind that this PR will be blocked until the dialog context is added, and then will have a re-review. Good job otherwise, I look forward to merging this in!

> Reviewed 2 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-:-NtWuvt2-K8y1kP04HI1:bv57l0a)* status: 2 of 5 files reviewed, 9 unresolved discussions (waiting on @caseycharleston and @Samathingamajig)

___
*[`src/pages/background/lib/deleteSchedule.ts` line 25 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWsxT04nyanJGMAMjE:-NtWsxT04nyanJGMAMjF:b34ntmp) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/deleteSchedule.ts#L25)):*
> ```XML
>     if (scheduleIndex < activeIndex) {
>         await UserScheduleStore.set('activeIndex', activeIndex - 1);
>     }
> ```

what happens if the currently active schedule is the last element? Then the active schedule would be nonexistent.



_Code quote:_
```XML
    if (scheduleIndex < activeIndex) {
        await UserScheduleStore.set('activeIndex', activeIndex - 1);
    }
```
___
*[`src/pages/background/lib/handleDuplicate.ts` line 11 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWt96Y38QbNj8HQAb5:-NtWt96Y38QbNj8HQAb6:b-xun63y) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L11)):*
> ```XML
>  */
> export default async function handleDuplicate(scheduleName: string): Promise<string> {
>     const schedules = await UserScheduleStore.get('schedules');
> ```

Can we remove the await call here, and pass in `schedules` as an argument to the function?
___
*[`src/pages/background/lib/handleDuplicate.ts` line 19 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWtkfS1gXuCpaDycFp:-NtWtkfS1gXuCpaDycFq:batozyc) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L19)):*
> ```XML
> 
>     // Regex matches strings of the form: " (int)"
>     const regex = / \((\d+)\)[^(]*$/;
> ```

This regex seems off, it will match " (123) abcdefdlfkjsdflkj" as well.

Instead, perhaps it should be `/\s+\((\d+)\)\s*$/`?



_Code quote:_
```XML
/ \((\d+)\)[^(]*$/
```
___
*[`src/pages/background/lib/handleDuplicate.ts` line 24 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#gh-1533197302:-NtWuMxN95dVtIH3dkG1:bbohma4) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L24)):*
<details><summary><i>Previously, Samathingamajig (Samuel Gunter) wrote…</i></summary><blockquote>

`baseName` can be `const`, and we don't need to wrap something that's ready a string in a template literal
```suggestion
    // Trim ending number
    const baseName = scheduleName.replace(regex, '');
    let newName = baseName;
```
</blockquote></details>

Additionally, I would recommend having the baseName be a group within the regex which you can extract, rather than replacing it away. Setting the initial index from the matcher would also be good, as it should only ever increment.


<!-- Sent from Reviewable.io -->

Samathingamajig reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)* status: 2 of 5 files reviewed, 9 unresolved discussions (waiting on @caseycharleston and @Razboy20)

___
*[`src/pages/background/lib/deleteSchedule.ts` line 25 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWsxT04nyanJGMAMjE:-NtX0UOL7-YZic0zo_g3:b-cmyvyv) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/deleteSchedule.ts#L25)):*
<details><summary><i>Previously, Razboy20 wrote…</i></summary><blockquote>

what happens if the currently active schedule is the last element? Then the active schedule would be nonexistent.
</blockquote></details>

i think that would only apply if the active schedule could be deleted, which for some reason it is not (the only justification i can think for that decision was we never want to be in a state with no schedules at all... but then why not just make a new blank schedule if that's ever the case?)


<!-- Sent from Reviewable.io -->

Razboy20 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)* status: 2 of 5 files reviewed, 9 unresolved discussions (waiting on @caseycharleston and @Samathingamajig)

___
*[`src/pages/background/lib/deleteSchedule.ts` line 25 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWsxT04nyanJGMAMjE:-NtX6PZ0CDZEkNjjQpoP:bj7avcc) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/deleteSchedule.ts#L25)):*
<details><summary><i>Previously, Samathingamajig (Samuel Gunter) wrote…</i></summary><blockquote>

i think that would only apply if the active schedule could be deleted, which for some reason it is not (the only justification i can think for that decision was we never want to be in a state with no schedules at all... but then why not just make a new blank schedule if that's ever the case?)
</blockquote></details>

You can actually delete the active schedule, that's what this code is referencing.


<!-- Sent from Reviewable.io -->

Razboy20 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)* status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @caseycharleston and @Samathingamajig)

___
*[`src/pages/background/lib/handleDuplicate.ts` line 0 at r2](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtXFYF6E-68VGXE_iLJ:-NtXFYF7ErRYA7g3A5dP:bzhgkmh) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/ac82ef92887163029b6f97492e2b51a22b36d610/src/pages/background/lib/handleDuplicate.ts#L0)):*
Currently, `handleDuplicate` isn't actually cloning the courses.


<!-- Sent from Reviewable.io -->

Samathingamajig reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)* status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @caseycharleston and @Razboy20)

___
*[`src/pages/background/lib/deleteSchedule.ts` line 25 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWsxT04nyanJGMAMjE:-NtXTGOj4iTJeN7kv710:b-uuo59y) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/deleteSchedule.ts#L25)):*
<details><summary><i>Previously, Razboy20 wrote…</i></summary><blockquote>

You can actually delete the active schedule, that's what this code is referencing.
</blockquote></details>

right above these lines are

```tsx
    if (scheduleIndex === activeIndex) {
        return 'Cannot delete active schedule';
    }
```

additionally,
```tsx
    const activeDeleteTitle = <Text>Invalid action!</Text>;
    const activeDeleteContent = <Text>Deleting the active schedule is disallowed.</Text>;
```

with the current state of the code, you cannot delete the active schedule

this section of code is referencing that if I'm on index 5, and index 2 is deleted, then the old index 5 is now index 4; if I'm on index 5, and index 8 is deleted, then the old index 5 is still index 5.


<!-- Sent from Reviewable.io -->

Razboy20 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)* status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @caseycharleston and @Samathingamajig)

___
*[`src/pages/background/lib/deleteSchedule.ts` line 25 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWsxT04nyanJGMAMjE:-NtXYPS6-WGrnyeG5E3e:b-jwnfw7) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/deleteSchedule.ts#L25)):*
<details><summary><i>Previously, Samathingamajig (Samuel Gunter) wrote…</i></summary><blockquote>

right above these lines are

```tsx
    if (scheduleIndex === activeIndex) {
        return 'Cannot delete active schedule';
    }
```

additionally,
```tsx
    const activeDeleteTitle = <Text>Invalid action!</Text>;
    const activeDeleteContent = <Text>Deleting the active schedule is disallowed.</Text>;
```

with the current state of the code, you cannot delete the active schedule

this section of code is referencing that if I'm on index 5, and index 2 is deleted, then the old index 5 is now index 4; if I'm on index 5, and index 8 is deleted, then the old index 5 is still index 5.
</blockquote></details>

Ah whoops, misread. In the future, I do want to be able to delete active schedules however, and it should just switch over to another one.


<!-- Sent from Reviewable.io -->

doprz commented:
Let's rebase (or update PR) with main to incorporate the new changes such as #168

caseycharleston commented:
> what happens if the currently active schedule is the last element? Then the active schedule would be nonexistent.
> 
> _Code quote:_
> 
> ```
>     if (scheduleIndex < activeIndex) {
>         await UserScheduleStore.set('activeIndex', activeIndex - 1);
>     }
> ```

I don't believe that is necessarily true, if we have a list of two elements and activeIndex = 1, deleting schedule 0 would then make the activeIndex = 0, which reflects the new position of the active schedule after a schedule is removed.

> _[`src/pages/background/lib/handleDuplicate.ts` line 11 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWt96Y38QbNj8HQAb5:-NtWt96Y38QbNj8HQAb6:b-xun63y) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L11)):_


> 
> > ```
> >  */
> > export default async function handleDuplicate(scheduleName: string): Promise<string> {
> >     const schedules = await UserScheduleStore.get('schedules');
> > ```
> 
> Can we remove the await call here, and pass in `schedules` as an argument to the function?

I originally tried this but I had issues using the schedules value because it is wrapped in a Serialized type. Not too sure how to use Serialized values correctly, may need some advice on that.
> 
> _[`src/pages/background/lib/handleDuplicate.ts` line 19 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtWtkfS1gXuCpaDycFp:-NtWtkfS1gXuCpaDycFq:batozyc) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L19)):_
> 
> > ```
> >     // Regex matches strings of the form: " (int)"
> >     const regex = / \((\d+)\)[^(]*$/;
> > ```
> 
> This regex seems off, it will match " (123) abcdefdlfkjsdflkj" as well.
> 
> Instead, perhaps it should be `/\s+\((\d+)\)\s*$/`?
> 
> _Code quote:_
> 
> ```
> / \((\d+)\)[^(]*$/
> ```

Changed to reflect your regex.

> 
> _[`src/pages/background/lib/handleDuplicate.ts` line 24 at r1](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#gh-1533197302:-NtWuMxN95dVtIH3dkG1:bbohma4) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/5c5c04aa40907ab20d6610049318514d8c65ec98/src/pages/background/lib/handleDuplicate.ts#L24)):_
> 
> Previously, Samathingamajig (Samuel Gunter) wrote…
> Additionally, I would recommend having the baseName be a group within the regex which you can extract, rather than replacing it away. Setting the initial index from the matcher would also be good, as it should only ever increment.

Changed the regex to have both a base name and index. I thought about setting the initial index to the one found by the matcher but what if we have:

add ->duplicate schedule (1)
add -> duplicate schedule (2)
delete -> duplicate schedule (1)
duplicate -> duplicate schedule (2)

If we only ever incremented, we would get duplicate schedule (2) even though 1 is an available number.

caseycharleston commented:
> > _[Reviewable](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182)_ status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @caseycharleston and @Samathingamajig)
> 
> _[`src/pages/background/lib/handleDuplicate.ts` line 0 at r2](https://reviewable.io/reviews/Longhorn-Developers/UT-Registration-Plus/182#-NtXFYF6E-68VGXE_iLJ:-NtXFYF7ErRYA7g3A5dP:bzhgkmh) ([raw file](https://github.com/Longhorn-Developers/UT-Registration-Plus/blob/ac82ef92887163029b6f97492e2b51a22b36d610/src/pages/background/lib/handleDuplicate.ts#L0)):_ Currently, `handleDuplicate` isn't actually cloning the courses.

Fixed in d370f0524e0229bd4c5e32400337080d58244ea4

caseycharleston commented:
Closed PR due to rebase issues on my part. see #207 for an up-to-date version of this PR.
